### PR TITLE
feat(Odin): added timepicker on overview [YTFRONT-2733]

### PR DESCRIPTION
**Pull request by mcf11y**
https://nda.ya.ru/t/F57IC79475siLM

Снимок экрана 2024-04-05 в 13 23 12

### CODE REVIEW

File: packages/ui/src/ui/pages/odin/controls/OdinOverview.tsx [hunk 1]
```diff
@@ -229,67 +267,196 @@ function OdinOverview(props: OdinOverviewProps) {
 
     const dispatch = useDispatch();
 
+    const isTimeCloseToCurrent = (time: number) => moment().diff(moment(time), 'minutes') < 5;
+
+    const isPickerTimeValid = (time: number) => {
+        return time < new Date().getTime() - HALF_HOUR;
+    };
+
+    const onPrevDateClick = () => {
```
Comment (ma-efremoff) [inline packages/ui/src/ui/pages/odin/controls/OdinOverview.tsx:276]: It is better to move all the login to actions and leave here simple callbacks like:
```ts
const onPrevDateClick = () => {
    dispatch(fetchPrevOdinOverview(cluster));
}

const onNextDateClick = () => {
    dispatch(fetchNextOdinOverview(cluster));
}

const onUpdate = (value) => {
    dispatch(setOdinOverivewFromTime(value));
}
```

File: packages/ui/src/ui/pages/odin/controls/OdinOverview.tsx [hunk 2]
```diff
@@ -36,29 +46,57 @@ import Link from '../../../components/Link/Link';
...
+        clearInterval(pollingIntervalId);
+        pollingIntervalId = null;
+    }
+}
+
 function useOdinOverviewDataLoader(cluster: string) {
     const dispatch = useDispatch();
+    const searchParams = useSearchParams();
+
     React.useEffect(() => {
```
Comment (ma-efremoff) [inline packages/ui/src/ui/pages/odin/controls/OdinOverview.tsx:85]: The body of the effect should not be changed

Comment (mcf11y) [reply packages/ui/src/ui/pages/odin/controls/OdinOverview.tsx:86]: deleted

File: packages/ui/src/ui/pages/odin/controls/OdinOverview.tsx (hunk 1 above)
Comment (mcf11y) [reply packages/ui/src/ui/pages/odin/controls/OdinOverview.tsx:277]: good, fixed

File: packages/ui/src/ui/hooks/use-search-params.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    set: never;
+    delete: never;
+    sort: never;
+}
+
+export function useSearchParams() {
+    const {search} = useLocation();
+
+    return useMemo(() => new URLSearchParams(search) as ReadOnlyURLSearchParams, [search]);
+}
```
[the remark this answers is not in the document]
Comment (mcf11y) [reply packages/ui/src/ui/hooks/use-search-params.ts:15]: fixed

File: packages/ui/src/ui/pages/odin/_actions/odin-overview.ts [hunk 4]
```diff
@@ -31,15 +32,23 @@ import {OdinRootState} from '../_reducers';
...
+        const oldCluster = getClusterHelper(getState);
+
+        dispatch({
+            type: ODIN_OVERVIEW_REQUEST,
+        });
+
+        if (oldCluster !== cluster) {
             const data = getOdinOverviewData(getState());
             _forEach(data, ({cancel}) => {
                 if (cancel) {
```
Comment (ma-efremoff) [inline packages/ui/src/ui/pages/odin/_actions/odin-overview.ts:54]: `cancel()` should be called anyway

File: packages/ui/src/ui/pages/odin/_actions/odin-overview.ts [hunk 5]
```diff
@@ -83,11 +105,49 @@ export function fetchOdinOverview(cluster: string): OdinOverviewThunkAction {
     };
 }
 
+export function fetchPrevOdinOverview(): OdinOverviewThunkAction {
+    return (dispatch, getState) => {
+        const state = getState();
+        const timeFrom = getOdinOverviewTimeFrom(state);
+        const timeTo = getOdinOverviewTimeTo(state);
```
Comment (ma-efremoff) [inline packages/ui/src/ui/pages/odin/_actions/odin-overview.ts:112]: why do we need to have `timeTo` variable?

File: packages/ui/src/ui/pages/odin/_actions/odin-overview.ts [hunk 6]
```diff
@@ -83,11 +105,49 @@ export function fetchOdinOverview(cluster: string): OdinOverviewThunkAction {
...
+
+        dispatch(fetchOdinOverview(cluster, timeFrom - OFFSET_IN_MINUTES));
+    };
+}
+
+export function fetchNextOdinOverview(): OdinOverviewThunkAction {
+    return (dispatch, getState) => {
+        const state = getState();
+        const timeFrom = getOdinOverviewTimeFrom(state);
+        const timeTo = getOdinOverviewTimeTo(state);
```
Comment (ma-efremoff) [inline packages/ui/src/ui/pages/odin/_actions/odin-overview.ts:127]: why do we need to have timeTo variable?

File: packages/ui/src/ui/pages/odin/_reducers/url-mapping.ts [hunk 7]
```diff
@@ -36,28 +37,35 @@ export const odinParams: LocationParameters = {
...
         stateKey: 'odin.details.odinCluster',
-        initialState: initialState.odinCluster,
+        initialState: overviewInitialState.dataCluster,
+    },
+    date: {
+        stateKey: 'odin.overview.timeFrom',
+        initialState: overviewInitialState.timeFrom,
+        options: {
+            parse: (timeFromUrl: string) => Number(timeFromUrl),
+        },
```
Comment (ma-efremoff) [inline packages/ui/src/ui/pages/odin/_reducers/url-mapping.ts:68]:
```ts
type: 'number'
```

**Review by ma-efremoff — APPROVED** [review]

**Pull request merged**